sprockets_statsd/statsd.py

```diff
@@ -0,0 +1,435 @@
+
+    .. attribute:: queue
+       :type: asyncio.Queue
+
+       Formatted metric payloads to send to the statsd server.  Enqueue
+       payloads to send them to the server.
+
+    .. attribute:: running
+       :type: asyncio.Event
+
+       Is the background task currently running?  This is the event that
+       :meth:`.run` sets when it starts and it remains set until the task
+       exits.
+
+    .. attribute:: stopped
+       :type: asyncio.Event
+
+       Is the background task currently stopped?  This is the event that
+       :meth:`.run` sets when it exits and that :meth:`.stop` blocks on
+       until the task stops.
+
+    """
+
+    protocol: typing.Union[StatsdProtocol, None]
+    _create_transport: typing.Callable[[], typing.Coroutine[
+        typing.Any, typing.Any, typing.Tuple[asyncio.BaseTransport,
+                                             StatsdProtocol]]]
+
+    def __init__(self,
+                 *,
+                 host,
+                 port: int = 8125,
+                 reconnect_sleep: float = 1.0,
+                 ip_protocol: int = socket.IPPROTO_TCP,
+                 wait_timeout: float = 0.1):
+        super().__init__()
+        if not host:
+            raise RuntimeError('host must be set')
+        try:
+            port = int(port)
+            if not port or port < 1:
+                raise RuntimeError(
+                    f'port must be a positive integer: {port!r}')
+        except (TypeError, ValueError):
+            raise RuntimeError(f'port must be a positive integer: {port!r}')
+
+        transport_creators = {
+            socket.IPPROTO_TCP: self._create_tcp_transport,
+            socket.IPPROTO_UDP: self._create_udp_transport,
+        }
+        try:
+            self._create_transport = transport_creators[ip_protocol]
+        except KeyError:
+            raise RuntimeError(f'ip_protocol {ip_protocol} is not supported')
+
+        self.host = host
+        self.port = port
+        self._ip_protocol = ip_protocol
+        self._reconnect_sleep = reconnect_sleep
+        self._wait_timeout = wait_timeout
+
+        self.running = asyncio.Event()
+        self.stopped = asyncio.Event()
+        self.stopped.set()
+        self.logger = logging.getLogger(__package__).getChild('Processor')
+        self.should_terminate = False
+        self.protocol = None
+        self.queue = asyncio.Queue()
+
+        self._failed_sends = []
+
+    @property
+    def connected(self) -> bool:
+        """Is the processor connected?"""
+        if self.protocol is None:
+            return False
+        return self.protocol.connected.is_set()
+
+    async def run(self):
+        """Maintains the connection and processes metric payloads."""
+        self.running.set()
+        self.stopped.clear()
+        self.should_terminate = False
+        while not self.should_terminate:
+            try:
+                await self._connect_if_necessary()
+                if self.connected:
```

Comment:
You mean `self.connected.is_set()` and other places below?

Reply:
It refers to the `Processor.connected` property which checks `self.protocol.connected.is_set()`.  A little confusing but they are in different classes and have slightly different use cases.